src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp

```diff
@@ -273,3 +273,26 @@ bool QgsValueRelationFieldFormatter::expressionIsUsable( const QString &expressi
     return false;
   return true;
 }
+
+QgsVectorLayer *QgsValueRelationFieldFormatter::resolveLayer( const QVariantMap &config )
```

Comment:
Sorry, I'm a bit late to this conversation -- but couldn't this logic be moved to QgsMapLayerRef::resolveByIdOrNameOnly? I would propose that something like this would work:

```
  TYPE *resolveByIdOrNameOnly( const QgsProject *project )
  {
    // first try by matching by layer ID, or weakly by source, name and provider
    if ( resolveWeakly( project ) )
      return layer;

    // fallback to checking by name only
    if ( project && !name.isEmpty() )
    {
      const QList<QgsMapLayer *> layers = project->mapLayersByName( name );
      for ( QgsMapLayer *l : layers )
      {
        if ( TYPE *tl = qobject_cast< TYPE *>( l ) )
        {
            setLayer( tl );
            return tl;
        }
      }
    }
    return nullptr;
  }
```

This has a few advantages:
- keeps the logic inside QgsMapLayerRef, which is arguably where it belongs
- prefers to match layers with a matching source, so in the case of a project with two duplicate named layers, the one which matches the original layer source will be preferred

Reply:
(Also another potential benefit -- in future we could extend QgsMapLayerRef to allow it to load layers into the project automatically -- e.g. if the ref isn't resolved to a matching layer, then create a new one using the stored source and provider data, and add it to the project's map layer store BUT NOT the layer tree. This would allow use of layer sources without having to manually load them first into the project for use cases like this)

Reply:
And that "could allow" must be read as "should allow", along with adding a new section to project properties for layers associated with a project which aren't shown in the layer tree (e.g. support tables you don't want to expose to users)

Reply:
Sounds good.

Reply:
@nyalldawson 

Should I add these or is it designed to be used in other ways?

```
typedef _LayerRef<QgsVectorLayer> QgsVectorLayerRef;
typedef _LayerRef<QgsRasterLayer> QgsRasterLayerRef;
typedef _LayerRef<QgsMeshLayer> QgsMeshLayerRef;
typedef _LayerRef<QgsMeshLayer> QgsPluginLayerRef;
```

Reply:
@elpaso they already exist -- see e.g. qgsvectorlayerref.h